ws/sha1.ts

```diff
@@ -47,17 +50,20 @@ export class Sha1 {
 
     this.block = this.start = this.bytes = this.hBytes = 0;
     this.finalized = this.hashed = false;
-    this.first = true;
   }
 
-  update(data: string | ArrayBuffer) {
+  update(data: string | ArrayBuffer | ArrayBufferView) {
     if (this.finalized) {
-      return;
+      return this;
     }
     let message;
     let notString = typeof data !== "string";
-    if (notString && data instanceof ArrayBuffer) {
-      message = new Uint8Array(data);
+    if (notString) {
+      if (data instanceof ArrayBuffer) {
+        message = new Uint8Array(data);
+      } else if (ArrayBuffer.isView(data)) {
```

Comment:
There is no final `else` here, which while unlikely if the types are not correct at runtime, the logic could fail.  Likely there is a problem with type narrowing here, because the `notString` is not narrowing the type.  Because I am just looking at the diff, I would have questioned the original code on using the `let notString` not really seeing a reassignment going on, which means just adding `typeof data !== "string"` in the condition would narrow the type and allow the `else` condition to be final.

Also you could just consider a ternary operator.  If `ArrayBuffer.isView(data)` isn't true, what could `data` be anyways here?

Reply:
Yeah, good catch